File: algorithms/python/_653_TwoSumIV_InputIsA_BST/_653_TwoSumIV_InputIsA_BST_std.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution1:
    """
    @param: : the root of tree
    @param: : the target sum
    @return: two numbers from tree which sum is n
    """

    def twoSum(self, root, n):
        # write your code here

        def search(visited, val):
            node = visited
            while node:
                if node.val == val:
                    return visited != node
                elif node.val < val:
                    node = node.right
                else:
                    node = node.left
            return False

        def traval(node):
            if not node:
                return None

            if search(node, n - node.val):
                return [node.val, n - node.val]
            left = traval(node.left)
            if left:
                return left
            right = traval(node.right)
            if right:
                return right

            return None

        return traval(root)

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution2:
    """
    @param: : the root of tree
    @param: : the target sum
    @return: two numbers from tree which sum is n
    """

    def twoSum(self, root, n):
        # write your code here
        if root is None:
            return None
        table = {}
        queue = []
        queue.append(root)
        table[root.val] = 1
        while queue:
            node = queue.pop(0)
            if node.val in table:
                return [node.val, n-node.val]
            if node.left:
                queue.append(node.left)
                table[node.left.val] = 1
            if node.right:
                queue.append(node.right)
                table[node.right.val] = 1
        return None

File: algorithms/cpp/_653_TwoSumIV_InputIsA_BST/_653_TwoSumIV_InputIsA_BST_std.cpp

# include <iostream>
#include <vector>

using namespace std;

int main() {
    cout << "new file!" << endl;
    return 0;
}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    explicit TreeNode(int val) {
        this->val = val;
        this->left = NULL;
        this->right = NULL;
    }
};

class Solution1 {
public:
    /*
     * @param : the root of tree
     * @param : the target sum
     * @return: two number from tree witch sum is n
     */
    vector<int> twoSum(TreeNode *root, int n) {
        // write your code here
        root_node = root;
        return dfs(root, n);
    }

private:
    TreeNode *root_node = NULL;

    vector<int> dfs(TreeNode *root, int n) {
        vector<int> ret;
        if (!root)
            return ret;

        ret = dfs(root->left, n);
        if (!ret.empty())
            return ret;

        if (helper(root_node, n - root->val, root)) {
            ret.push_back(root->val);
            ret.push_back(n - root->val);
            return ret;
        }

        ret = dfs(root->right, n);
        return ret;
    }

    bool helper(TreeNode *root, int val, TreeNode *node) {
        if (!root)
            return false;

        if (root->val == val) {
            if (root != node)
                return true;
            else {
                return helper(root->left, val, node) || helper(root->right, val, node);
            }
        } else if (root->val < val)
            return helper(root->right, val, node);
        else
            return helper(root->left, val, node);
    }
    // vector<int> twoSum(TreeNode* root, int k) {
    //     vector<int> ret;
    //     if (!root) return ret;
    //     unordered_set<int> s;
    //     queue<TreeNode*> q{{root}};
    //     while (!q.empty()) {
    //       auto t = q.front(); q.pop();
    //       if (s.count(k - t->val)) {
    //           ret.push_back(t->val);
    //           ret.push_back(k - t->val);
    //           return ret;
    //       }
    //       s.insert(t->val);
    //       if (t->left) q.push(t->left);
    //       if (t->right) q.push(t->right);
    //     }
    //     return ret;
    // }
};

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution2 {
public:
    /*
     * @param : the root of tree
     * @param : the target sum
     * @return: two number from tree witch sum is n
     */
    vector<int> res;
    vector<int> twoSum(TreeNode * root, int n) {
        // write your code here
        if(root == NULL)
            return res;
        collect(root);
        // for(int i = 0; i < res.size(); i++)
        // {
        //     cout << res[i] << " ";
        // }
        //collect(root);
        return find(n);
        //return res;
    }

    void collect(TreeNode *root)
    {
        if(root != NULL)
        {
            collect(root->left);
            res.push_back(root->val);
            collect(root->right);
        }
    }
    vector<int> find(int target)
    {
        int start = 0;
        int end = res.size() - 1;
        vector<int> answer;
        while(start < end)
        {
            if(res[start] + res[end] < target)
            {
                start++;
            }
            else if(res[start] + res[end] > target)
            {
                end--;
            }
            else
            {
                answer.push_back(res[start]);
                answer.push_back(res[end]);
                return answer;
            }
        }
        return {};
    }
};